src/backend/distributed/executor/repartition_logic.c

```diff
@@ -0,0 +1,208 @@
+/*-------------------------------------------------------------------------
+ *
+ * repartition_logic.c
+ *
+ * This file contains repartition specific logic.
+ * ExecuteDependedTasks takes a list of top level tasks. Its logic is as follows:
+ * - It generates all the tasks by descending in the tasks tree. Note that each task
+ *  has a dependedTaskList.
+ * - It generates FetchTask queryStrings with the MapTask queries. It uses the first replicate to
+ *  fetch data when replication factor is > 1. Note that if a task fails in any replica adaptive executor
+ *  gives an error, so if we come to a fetchTask we know for sure that its dependedMapTask is executed in all
+ *  replicas.
+ * - It creates schemas in each worker in a single transaction to store intermediate results.
+ * - It iterates all tasks and finds the ones whose dependencies are already executed, and executes them with
+ *  adaptive executor logic.
+ *
+ *
+ * Repartition queries do not begin a transaction even if we are in
+ * a transaction block. As we dont begin a transaction, they wont see the
+ * DDLs that happened earlier in the transaction because we dont have that
+ * transaction id with repartition queries. Therefore we error in this case.
+ *
+ * Copyright (c) Citus Data, Inc.
+ */
+
+#include "postgres.h"
+#include "access/hash.h"
+#include "distributed/hash_helpers.h"
+
+#include "distributed/directed_acylic_graph_execution_logic.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/adaptive_executor.h"
+#include "distributed/worker_manager.h"
+#include "distributed/multi_server_executor.h"
+#include "distributed/repartition_logic.h"
+#include "distributed/worker_transaction.h"
+#include "distributed/worker_manager.h"
+#include "distributed/transaction_management.h"
+#include "distributed/multi_task_tracker_executor.h"
+#include "distributed/worker_transaction.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/listutils.h"
+#include "distributed/transmit.h"
+
+
+static List * ExtractMergeTasks(List *allTasks);
+static List * CreateTemporarySchemasForMergeTasks(List *mergeTasks);
+static List * ExtractJobIdsFromTasks(List *mergeTasks);
+static void CreateSchemasOnAllWorkers(char *createSchemasCommand);
+static char * GenerateCreateSchemasCommand(List *jobIds);
+static char * GenerateJobCommands(List *jobIds, char *templateCommand);
+static char * GenerateDeleteJobsCommand(List *jobIds);
+static void RemoveTempJobDirs(List *jobIds);
+
+
+/*
+ * ExecuteDependedTasks executes all tasks except the top level tasks
+ * in order from the task tree. At a time, it can execute different tasks from
+ * different jobs.
+ */
+void
+ExecuteDependedTasks(List *topLevelTasks)
+{
+	EnsureNoModificationsHaveBeenDone();
+
+	List *allTasks = TaskAndExecutionList(topLevelTasks);
+
+	List *mergeTasks = ExtractMergeTasks(allTasks);
```

Comment:
I think the point of this is ultimately just to get a set of job IDs. If we passed the Job to ExecuteDependedTasks. Would we be able to get the job IDs in a cheaper way?

Reply:
IIRC jobs do not contain any dependent jobs. Tasks contain dependent tasks. This can be fixed in the planner part I guess. Then we can get the job ids from by traversing the top job.

Reply:
I don't think that's true. The EXPLAIN code relies on the dependedJobList

Reply:
I guess you are right, I have debugged some queries and it seems that they have dependentJobList. I have changed it to use a DFS to find job ids starting from the top level job.